### PR TITLE
low level pyro training plan

### PR DESCRIPTION
**Pull request by adamgayoso**
Add a lower level Pyro training plan to allow use of vanilla torch optimization

### CODE REVIEW

**Comment by adamgayoso** [comment]
@vitkl @fritzo this should allow lightning to do it's thing with accelerators and work with pyro.

Note that this callback:

https://github.com/scverse/scvi-tools/blob/2816e3da080a2aae59b939bb55acb19964d87d63/scvi/model/base/_pyromixin.py#L18

would always need to be used to ensure `self.module.parameters()` is populated with everything before `configure_optimizers` is called. Note that the callback hook used might be too late in the code above.

**Comment by vitkl** [comment]
This is cool.

1. Does pyro work with torch rather than pyro-wrapped optimisers? This probably requires some changes to models, no?
2. `training_step()` contents replace what is normally done inside `svi.step()`, right?

> so it might need to occur in the setup stage.

How would you do this?

File: scvi/train/_trainingplans.py [hunk 1]
```diff
@@ -907,6 +907,102 @@ def kl_weight(self):
...
+        n_steps_kl_warmup: Union[int, None] = None,
+        n_epochs_kl_warmup: Union[int, None] = 400,
+        scale_elbo: float = 1.0,
+    ):
+        self.module = pyro_module
+        self._n_obs_training = None
+
+        optim_kwargs = optim_kwargs if isinstance(optim_kwargs, dict) else dict()
+        if "lr" not in optim_kwargs.keys():
+            optim_kwargs.update({"lr": 1e-3})
```
Comment (vitkl) [inline scvi/train/_trainingplans.py:953]: `optim_kwargs` should be an attribute, right?

```
self.optim_kwargs = optim_kwargs
```

File: scvi/train/_trainingplans.py [hunk 2]
```diff
@@ -907,6 +907,102 @@ def kl_weight(self):
...
+            self.scale_fn(self.module.model),
+            self.scale_fn(self.module.guide),
+            *args,
+            **kwargs,
+        )
+        return {"loss": loss}
+
+    def configure_optimizers(self):
+        """Configure optimizers for the model."""
+        return self.optim(self.module.parameters(), **self.optim_kwargs)
```
Comment (vitkl) [inline scvi/train/_trainingplans.py:992]: I get this warning

```python
/nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/core/optimizer.py:182: UserWarning: `LightningModule.configure_optimizers` returned `None`, this fit will run with no optimizer
```

Comment (adamgayoso) [reply scvi/train/_trainingplans.py:994]: Are you sure you're using it correctly? I don't see how it's None

Comment (vitkl) [reply scvi/train/_trainingplans.py:994]: True, I was accidentally providing pyro `pyro.optim.Adam`.

**Comment by vitkl** [comment]
The next error - for some reason guide parameters and the model inputs are not on the same device.

```python
File /nfs/team205/vk7/sanger_projects/my_packages/cell2state/cell2state/models/_programme_model.py:942, in ProgrammeModel.train(self, max_epochs, use_gpu, train_size, validation_size, batch_size, gene_batch_size, drop_last, early_stopping, lr, num_particles, scale_elbo, training_plan, plan_kwargs, dl_kwargs, simple_progress_bar, accumulate_grad_batches, **trainer_kwargs)
    933 print(trainer_kwargs)
    934 runner = TrainRunner(
    935     self,
    936     training_plan=training_plan,
   (...)
    940     **trainer_kwargs,
    941 )
--> 942 return runner()

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/scvi/train/_trainrunner.py:82, in TrainRunner.__call__(self)
     79 if hasattr(self.data_splitter, "n_val"):
     80     self.training_plan.n_obs_validation = self.data_splitter.n_val
---> 82 self.trainer.fit(self.training_plan, self.data_splitter)
     83 self._update_history()
     85 # data splitter only gets these attrs after fit

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/scvi/train/_trainer.py:188, in Trainer.fit(self, *args, **kwargs)
    182 if isinstance(args[0], PyroTrainingPlan):
    183     warnings.filterwarnings(
    184         action="ignore",
    185         category=UserWarning,
    186         message="`LightningModule.configure_optimizers` returned `None`",
    187     )
--> 188 super().fit(*args, **kwargs)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:696, in Trainer.fit(self, model, train_dataloaders, val_dataloaders, datamodule, ckpt_path)
    677 r"""
    678 Runs the full optimization routine.
    679 
   (...)
    693     datamodule: An instance of :class:`~pytorch_lightning.core.datamodule.LightningDataModule`.
    694 """
    695 self.strategy.model = model
--> 696 self._call_and_handle_interrupt(
    697     self._fit_impl, model, train_dataloaders, val_dataloaders, datamodule, ckpt_path
    698 )

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:650, in Trainer._call_and_handle_interrupt(self, trainer_fn, *args, **kwargs)
    648         return self.strategy.launcher.launch(trainer_fn, *args, trainer=self, **kwargs)
    649     else:
--> 650         return trainer_fn(*args, **kwargs)
    651 # TODO(awaelchli): Unify both exceptions below, where `KeyboardError` doesn't re-raise
    652 except KeyboardInterrupt as exception:

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:735, in Trainer._fit_impl(self, model, train_dataloaders, val_dataloaders, datamodule, ckpt_path)
    731 ckpt_path = ckpt_path or self.resume_from_checkpoint
    732 self._ckpt_path = self.__set_ckpt_path(
    733     ckpt_path, model_provided=True, model_connected=self.lightning_module is not None
    734 )
--> 735 results = self._run(model, ckpt_path=self.ckpt_path)
    737 assert self.state.stopped
    738 self.training = False

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:1166, in Trainer._run(self, model, ckpt_path)
   1162 self._checkpoint_connector.restore_training_state()
   1164 self._checkpoint_connector.resume_end()
-> 1166 results = self._run_stage()
   1168 log.detail(f"{self.__class__.__name__}: trainer tearing down")
   1169 self._teardown()

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:1252, in Trainer._run_stage(self)
   1250 if self.predicting:
   1251     return self._run_predict()
-> 1252 return self._run_train()

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:1283, in Trainer._run_train(self)
   1280 self.fit_loop.trainer = self
   1282 with torch.autograd.set_detect_anomaly(self._detect_anomaly):
-> 1283     self.fit_loop.run()

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/loop.py:200, in Loop.run(self, *args, **kwargs)
    198 try:
    199     self.on_advance_start(*args, **kwargs)
--> 200     self.advance(*args, **kwargs)
    201     self.on_advance_end()
    202     self._restarting = False

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/fit_loop.py:271, in FitLoop.advance(self)
    267 self._data_fetcher.setup(
    268     dataloader, batch_to_device=partial(self.trainer._call_strategy_hook, "batch_to_device", dataloader_idx=0)
    269 )
    270 with self.trainer.profiler.profile("run_training_epoch"):
--> 271     self._outputs = self.epoch_loop.run(self._data_fetcher)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/loop.py:200, in Loop.run(self, *args, **kwargs)
    198 try:
    199     self.on_advance_start(*args, **kwargs)
--> 200     self.advance(*args, **kwargs)
    201     self.on_advance_end()
    202     self._restarting = False

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py:203, in TrainingEpochLoop.advance(self, data_fetcher)
    200     self.batch_progress.increment_started()
    202     with self.trainer.profiler.profile("run_training_batch"):
--> 203         batch_output = self.batch_loop.run(kwargs)
    205 self.batch_progress.increment_processed()
    207 # update non-plateau LR schedulers
    208 # update epoch-interval ones only when we are at the end of training epoch

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/loop.py:200, in Loop.run(self, *args, **kwargs)
    198 try:
    199     self.on_advance_start(*args, **kwargs)
--> 200     self.advance(*args, **kwargs)
    201     self.on_advance_end()
    202     self._restarting = False

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/batch/training_batch_loop.py:87, in TrainingBatchLoop.advance(self, kwargs)
     83 if self.trainer.lightning_module.automatic_optimization:
     84     optimizers = _get_active_optimizers(
     85         self.trainer.optimizers, self.trainer.optimizer_frequencies, kwargs.get("batch_idx", 0)
     86     )
---> 87     outputs = self.optimizer_loop.run(optimizers, kwargs)
     88 else:
     89     outputs = self.manual_loop.run(kwargs)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/loop.py:200, in Loop.run(self, *args, **kwargs)
    198 try:
    199     self.on_advance_start(*args, **kwargs)
--> 200     self.advance(*args, **kwargs)
    201     self.on_advance_end()
    202     self._restarting = False

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:201, in OptimizerLoop.advance(self, optimizers, kwargs)
    198 def advance(self, optimizers: List[Tuple[int, Optimizer]], kwargs: OrderedDict) -> None:  # type: ignore[override]
    199     kwargs = self._build_kwargs(kwargs, self.optimizer_idx, self._hiddens)
--> 201     result = self._run_optimization(kwargs, self._optimizers[self.optim_progress.optimizer_position])
    202     if result.loss is not None:
    203         # automatic optimization assumes a loss needs to be returned for extras to be considered as the batch
    204         # would be skipped otherwise
    205         self._outputs[self.optimizer_idx] = result.asdict()

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:248, in OptimizerLoop._run_optimization(self, kwargs, optimizer)
    240         closure()
    242 # ------------------------------
    243 # BACKWARD PASS
    244 # ------------------------------
    245 # gradient update with accumulated gradients
    246 else:
    247     # the `batch_idx` is optional with inter-batch parallelism
--> 248     self._optimizer_step(optimizer, opt_idx, kwargs.get("batch_idx", 0), closure)
    250 result = closure.consume_result()
    252 if result.loss is not None:
    253     # if no result, user decided to skip optimization
    254     # otherwise update running loss + reset accumulated loss
    255     # TODO: find proper way to handle updating running loss

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:358, in OptimizerLoop._optimizer_step(self, optimizer, opt_idx, batch_idx, train_step_and_backward_closure)
    355     self.optim_progress.optimizer.step.increment_ready()
    357 # model hook
--> 358 self.trainer._call_lightning_module_hook(
    359     "optimizer_step",
    360     self.trainer.current_epoch,
    361     batch_idx,
    362     optimizer,
    363     opt_idx,
    364     train_step_and_backward_closure,
    365     on_tpu=isinstance(self.trainer.accelerator, TPUAccelerator),
    366     using_native_amp=(self.trainer.amp_backend == AMPType.NATIVE),
    367     using_lbfgs=is_lbfgs,
    368 )
    370 if not should_accumulate:
    371     self.optim_progress.optimizer.step.increment_completed()

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:1550, in Trainer._call_lightning_module_hook(self, hook_name, pl_module, *args, **kwargs)
   1547 pl_module._current_fx_name = hook_name
   1549 with self.profiler.profile(f"[LightningModule]{pl_module.__class__.__name__}.{hook_name}"):
-> 1550     output = fn(*args, **kwargs)
   1552 # restore current_fx when nested context
   1553 pl_module._current_fx_name = prev_fx_name

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/core/module.py:1705, in LightningModule.optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx, optimizer_closure, on_tpu, using_native_amp, using_lbfgs)
   1623 def optimizer_step(
   1624     self,
   1625     epoch: int,
   (...)
   1632     using_lbfgs: bool = False,
   1633 ) -> None:
   1634     r"""
   1635     Override this method to adjust the default way the :class:`~pytorch_lightning.trainer.trainer.Trainer` calls
   1636     each optimizer.
   (...)
   1703 
   1704     """
-> 1705     optimizer.step(closure=optimizer_closure)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/core/optimizer.py:168, in LightningOptimizer.step(self, closure, **kwargs)
    165     raise MisconfigurationException("When `optimizer.step(closure)` is called, the closure should be callable")
    167 assert self._strategy is not None
--> 168 step_output = self._strategy.optimizer_step(self._optimizer, self._optimizer_idx, closure, **kwargs)
    170 self._on_after_step()
    172 return step_output

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/strategies/strategy.py:216, in Strategy.optimizer_step(self, optimizer, opt_idx, closure, model, **kwargs)
    206 """Performs the actual optimizer step.
    207 
    208 Args:
   (...)
    213     **kwargs: Any extra arguments to ``optimizer.step``
    214 """
    215 model = model or self.lightning_module
--> 216 return self.precision_plugin.optimizer_step(model, optimizer, opt_idx, closure, **kwargs)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py:153, in PrecisionPlugin.optimizer_step(self, model, optimizer, optimizer_idx, closure, **kwargs)
    151 if isinstance(model, pl.LightningModule):
    152     closure = partial(self._wrap_closure, model, optimizer, optimizer_idx, closure)
--> 153 return optimizer.step(closure=closure, **kwargs)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/torch/optim/optimizer.py:88, in Optimizer._hook_for_profile.<locals>.profile_hook_step.<locals>.wrapper(*args, **kwargs)
     86 profile_name = "Optimizer.step#{}.step".format(obj.__class__.__name__)
     87 with torch.autograd.profiler.record_function(profile_name):
---> 88     return func(*args, **kwargs)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/torch/autograd/grad_mode.py:27, in _DecoratorContextManager.__call__.<locals>.decorate_context(*args, **kwargs)
     24 @functools.wraps(func)
     25 def decorate_context(*args, **kwargs):
     26     with self.clone():
---> 27         return func(*args, **kwargs)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/torch/optim/adam.py:100, in Adam.step(self, closure)
     98 if closure is not None:
     99     with torch.enable_grad():
--> 100         loss = closure()
    102 for group in self.param_groups:
    103     params_with_grad = []

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py:138, in PrecisionPlugin._wrap_closure(self, model, optimizer, optimizer_idx, closure)
    125 def _wrap_closure(
    126     self,
    127     model: "pl.LightningModule",
   (...)
    130     closure: Callable[[], Any],
    131 ) -> Any:
    132     """This double-closure allows makes sure the ``closure`` is executed before the
    133     ``on_before_optimizer_step`` hook is called.
    134 
    135     The closure (generally) runs ``backward`` so this allows inspecting gradients in this hook. This structure is
    136     consistent with the ``PrecisionPlugin`` subclasses that cannot pass ``optimizer.step(closure)`` directly.
    137     """
--> 138     closure_result = closure()
    139     self._after_closure(model, optimizer, optimizer_idx)
    140     return closure_result

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:146, in Closure.__call__(self, *args, **kwargs)
    145 def __call__(self, *args: Any, **kwargs: Any) -> Optional[Tensor]:
--> 146     self._result = self.closure(*args, **kwargs)
    147     return self._result.loss

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:132, in Closure.closure(self, *args, **kwargs)
    131 def closure(self, *args: Any, **kwargs: Any) -> ClosureResult:
--> 132     step_output = self._step_fn()
    134     if step_output.closure_loss is None:
    135         self.warning_cache.warn("`training_step` returned `None`. If this was on purpose, ignore this warning...")

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:407, in OptimizerLoop._training_step(self, kwargs)
    398 """Performs the actual train step with the tied hooks.
    399 
    400 Args:
   (...)
    404     A ``ClosureResult`` containing the training step output.
    405 """
    406 # manually capture logged metrics
--> 407 training_step_output = self.trainer._call_strategy_hook("training_step", *kwargs.values())
    408 self.trainer.strategy.post_training_step()
    410 model_output = self.trainer._call_lightning_module_hook("training_step_end", training_step_output)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:1704, in Trainer._call_strategy_hook(self, hook_name, *args, **kwargs)
   1701     return
   1703 with self.profiler.profile(f"[Strategy]{self.strategy.__class__.__name__}.{hook_name}"):
-> 1704     output = fn(*args, **kwargs)
   1706 # restore current_fx when nested context
   1707 pl_module._current_fx_name = prev_fx_name

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pytorch_lightning/strategies/strategy.py:358, in Strategy.training_step(self, *args, **kwargs)
    356 with self.precision_plugin.train_step_context():
    357     assert isinstance(self.model, TrainingStep)
--> 358     return self.model.training_step(*args, **kwargs)

File /nfs/team205/vk7/sanger_projects/my_packages/cell2state/cell2state/models/_programme_model.py:134, in LowLevelPyroTrainingPlan.training_step(self, batch, batch_idx)
    132     kwargs.update({"kl_weight": self.kl_weight})
    133 # pytorch lightning requires a Tensor object for loss
--> 134 loss = self.differentiable_loss_fn(
    135     self.scale_fn(self.module.model),
    136     self.scale_fn(self.module.guide),
    137     *args,
    138     **kwargs,
    139 )
    140 return {"loss": loss}

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/infer/trace_elbo.py:121, in Trace_ELBO.differentiable_loss(self, model, guide, *args, **kwargs)
    119 loss = 0.0
    120 surrogate_loss = 0.0
--> 121 for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):
    122     loss_particle, surrogate_loss_particle = self._differentiable_loss_particle(
    123         model_trace, guide_trace
    124     )
    125     surrogate_loss += surrogate_loss_particle / self.num_particles

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/infer/elbo.py:236, in ELBO._get_traces(self, model, guide, args, kwargs)
    234 else:
    235     for i in range(self.num_particles):
--> 236         yield self._get_trace(model, guide, args, kwargs)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/infer/trace_elbo.py:57, in Trace_ELBO._get_trace(self, model, guide, args, kwargs)
     52 def _get_trace(self, model, guide, args, kwargs):
     53     """
     54     Returns a single trace from the guide, and the model that is run
     55     against it.
     56     """
---> 57     model_trace, guide_trace = get_importance_trace(
     58         "flat", self.max_plate_nesting, model, guide, args, kwargs
     59     )
     60     if is_validation_enabled():
     61         check_if_enumerated(guide_trace)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/infer/enum.py:57, in get_importance_trace(graph_type, max_plate_nesting, model, guide, args, kwargs, detach)
     55     if detach:
     56         raise NotImplementedError("GuideMessenger does not support detach")
---> 57     guide(*args, **kwargs)
     58     model_trace, guide_trace = unwrapped_guide.get_traces()
     59 else:

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/poutine/messenger.py:12, in _context_wrap(context, fn, *args, **kwargs)
     10 def _context_wrap(context, fn, *args, **kwargs):
     11     with context:
---> 12         return fn(*args, **kwargs)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/nn/module.py:677, in pyro_method.<locals>.cached_fn(self, *args, **kwargs)
    674 @functools.wraps(fn)
    675 def cached_fn(self, *args, **kwargs):
    676     with self._pyro_context:
--> 677         return fn(self, *args, **kwargs)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/infer/autoguide/effect.py:46, in AutoMessenger.__call__(self, *args, **kwargs)
     44 self._outer_plates = tuple(f.name for f in get_plates())
     45 try:
---> 46     return super().__call__(*args, **kwargs)
     47 finally:
     48     del self._outer_plates

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/poutine/guide.py:45, in GuideMessenger.__call__(self, *args, **kwargs)
     43 try:
     44     with self:
---> 45         self.model(*args, **kwargs)
     46 finally:
     47     del self.args_kwargs

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/nn/module.py:448, in PyroModule.__call__(self, *args, **kwargs)
    446 def __call__(self, *args, **kwargs):
    447     with self._pyro_context:
--> 448         result = super().__call__(*args, **kwargs)
    449     if (
    450         pyro.settings.get("validate_poutine")
    451         and not self._pyro_context.active
    452         and _is_module_local_param_enabled()
    453     ):
    454         self._check_module_local_param_usage()

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/torch/nn/modules/module.py:1110, in Module._call_impl(self, *input, **kwargs)
   1106 # If we don't have any hooks, we want to skip the rest of the logic in
   1107 # this function, and just call forward.
   1108 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1109         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1110     return forward_call(*input, **kwargs)
   1111 # Do not call functions when jit is used
   1112 full_backward_hooks, non_full_backward_hooks = [], []

File /nfs/team205/vk7/sanger_projects/my_packages/cell2state/cell2state/models/cell2state/_module_cell2state.py:1899, in MultimodalCell2StatePyroModel.forward(self, x_data, idx, batch_index, rna_index, chr_index, extra_categoricals, var_categoricals, gene_bool, local_indices, region_motif_coo, tf_bool, variables_mask, gene_region_coo, region_region_coo, tn5_bias, free_floating_tt, tf_rna_rate_w_ch, cell_offset)
   1891         deep_setattr(
   1892             self.weights,
   1893             name,
   (...)
   1896             ).to(region_motif_coo_rhp.device),
   1897         )
   1898     module = deep_getattr(self.weights, name)
-> 1899     region_motif_coo_rhp = module(region_motif_coo_rhp.unsqueeze(-3))
   1900 # Switch between minibatch genes and minibatch cells (standard) modes ==========================
   1901 if self.minibatch_genes:
   1902     # Extract parameters when training in minibatches of genes

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/nn/module.py:448, in PyroModule.__call__(self, *args, **kwargs)
    446 def __call__(self, *args, **kwargs):
    447     with self._pyro_context:
--> 448         result = super().__call__(*args, **kwargs)
    449     if (
    450         pyro.settings.get("validate_poutine")
    451         and not self._pyro_context.active
    452         and _is_module_local_param_enabled()
    453     ):
    454         self._check_module_local_param_usage()

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/torch/nn/modules/module.py:1110, in Module._call_impl(self, *input, **kwargs)
   1106 # If we don't have any hooks, we want to skip the rest of the logic in
   1107 # this function, and just call forward.
   1108 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1109         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1110     return forward_call(*input, **kwargs)
   1111 # Do not call functions when jit is used
   1112 full_backward_hooks, non_full_backward_hooks = [], []

File /nfs/team205/vk7/sanger_projects/my_packages/cell2state/cell2state/nn/TFaffinityScanner.py:150, in TFaffinityScanner.forward(self, x)
    148 # Motif weights
    149 weights_name = "motif_loc"
--> 150 if getattr(self.weights, weights_name, None) is None:
    151     deep_setattr(
    152         self.weights,
    153         weights_name,
   (...)
    161         ),
    162     )
    163 motif_loc = deep_getattr(self.weights, weights_name)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/nn/module.py:520, in PyroModule.__getattr__(self, name)
    518     if not hasattr(prior, "sample"):  # if not a distribution
    519         prior = prior(self)
--> 520     value = pyro.sample(fullname, prior)
    521     context.set(fullname, value)
    522 return value

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/primitives.py:163, in sample(name, fn, *args, **kwargs)
    146 msg = {
    147     "type": "sample",
    148     "name": name,
   (...)
    160     "continuation": None,
    161 }
    162 # apply the stack and return its return value
--> 163 apply_stack(msg)
    164 return msg["value"]

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/poutine/runtime.py:213, in apply_stack(initial_msg)
    209 for frame in reversed(stack):
    211     pointer = pointer + 1
--> 213     frame._process_message(msg)
    215     if msg["stop"]:
    216         break

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/poutine/messenger.py:162, in Messenger._process_message(self, msg)
    160 method = getattr(self, "_pyro_{}".format(msg["type"]), None)
    161 if method is not None:
--> 162     return method(msg)
    163 return None

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/poutine/guide.py:62, in GuideMessenger._pyro_sample(self, msg)
     60 prior = msg["fn"]
     61 msg["infer"]["prior"] = prior
---> 62 posterior = self.get_posterior(msg["name"], prior)
     63 if isinstance(posterior, torch.Tensor):
     64     posterior = dist.Delta(posterior, event_dim=prior.event_dim)

File /nfs/team283/vk7/software/miniconda3farm5/envs/test_scvi19_cuda113/lib/python3.9/site-packages/pyro/infer/autoguide/effect.py:278, in AutoHierarchicalNormalMessenger.get_posterior(self, name, prior)
    275 if (self._hierarchical_sites is None) or (name in self._hierarchical_sites):
    276     # If hierarchical_sites not specified all sites are assumed to be hierarchical
    277     loc, scale, weight = self._get_params(name, prior)
--> 278     loc = loc + transform.inv(prior.mean) * weight
    279     posterior = dist.TransformedDistribution(
    280         dist.Normal(loc, scale).to_event(transform.domain.event_dim),
    281         transform.with_cache(),
    282     )
    283     return posterior

RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```

File: scvi/train/_trainingplans.py [hunk 3]
```diff
@@ -847,9 +878,69 @@ def n_obs_training(self, n_obs: int):
...
+        self.automatic_optimization = False
+
+        self.svi = pyro.infer.SVI(
+            model=self.scale_fn(self.module.model),
+            guide=self.scale_fn(self.module.guide),
+            optim=self.optim,
+            loss=self.loss_fn,
+        )
+
+        self._dummy_param = torch.nn.Parameter(torch.Tensor([0.0]))
```
Comment (vitkl) [inline scvi/train/_trainingplans.py:943]: What is the purpose of this?

File: scvi/train/_trainingplans.py [hunk 4]
```diff
@@ -847,9 +878,69 @@ def n_obs_training(self, n_obs: int):
...
+            scale_elbo=scale_elbo,
+        )
+        optim_kwargs = optim_kwargs if isinstance(optim_kwargs, dict) else dict()
+        if "lr" not in optim_kwargs.keys():
+            optim_kwargs.update({"lr": 1e-3})
+        self.optim = (
+            pyro.optim.Adam(optim_args=optim_kwargs) if optim is None else optim
+        )
+
+        self.automatic_optimization = False
```
Comment (vitkl) [inline scvi/train/_trainingplans.py:934]: Would be good to add a comment about what this means

File: scvi/model/base/_pyromixin.py [hunk 5]
```diff
@@ -46,6 +46,41 @@ def on_train_start(self, trainer, pl_module):
             break
 
 
+class PyroModelGuideWarmup(Callback):
```
Comment (vitkl) [inline scvi/model/base/_pyromixin.py:49]: Which test would fail if this doesn't work?

File: tests/models/test_pyro.py [hunk 6]
```diff
@@ -183,6 +184,32 @@ def _create_indices_adata_manager(adata: AnnData) -> AnnDataManager:
...
+    adata_manager = _create_indices_adata_manager(adata)
+    train_dl = AnnDataLoader(adata_manager, shuffle=True, batch_size=128)
+    pyro.clear_param_store()
+    model = BayesianRegressionModule(in_features=adata.shape[1], out_features=1)
+    plan = LowLevelPyroTrainingPlan(model)
+    plan.n_obs_training = len(train_dl.indices)
+    trainer = Trainer(
+        gpus=use_gpu,
+        max_epochs=2,
+        callbacks=[PyroModelGuideWarmup(train_dl)],
```
Comment (vitkl) [inline tests/models/test_pyro.py:199]: Is it essential to provide a dataloader like `PyroModelGuideWarmup(train_dl)`?

File: scvi/model/base/_pyromixin.py (hunk 5 above)
Comment (adamgayoso) [reply scvi/model/base/_pyromixin.py:49]: low level test

File: scvi/train/_trainingplans.py (hunk 4 above)
Comment (adamgayoso) [reply scvi/train/_trainingplans.py:934]: done

File: tests/models/test_pyro.py (hunk 6 above)
Comment (adamgayoso) [reply tests/models/test_pyro.py:199]: yes

Comment (adamgayoso) [reply tests/models/test_pyro.py:199]: actually no it's not required

**Comment by fritzo** [comment]
unrelated to this PR, but related to comments, I'd like to evangelize github markdown's [ `<details>` ](https://gist.github.com/ericclemmons/b146fe5da72ca1f706b2ef72a20ac39d) tag for pasting stack traces into PR comments 😉 The syntax looks like this:

```
<details>
                     # <-- you seem to need a newline here
long
list
of
stuff
                       # <-- and a newline here
</details>
```
and the results look like this

<details>

```
long
list
of
stuff
```

</details>

**Pull request merged**